onnxruntime/core/optimizer/transpose_optimization/onnx_transpose_optimization.cc

```diff
@@ -245,32 +245,232 @@ static std::unique_ptr<api::NodeRef> GetDQIfProducingValue(const api::GraphRef&
                                                                                      : std::unique_ptr<api::NodeRef>();
 }
 
+// Forward declarations.
+static bool NormalizeAndValidateAxes(std::vector<int64_t>& axes, size_t rank);
+static std::optional<std::vector<int64_t>> ReadFromAttrOrInput(const api::GraphRef& graph, api::NodeRef& node,
+                                                               std::string_view attr_name, size_t inp_index,
+                                                               int64_t opset);
+static int64_t UnsqueezeAxis(gsl::span<const int64_t> sorted_positive_unsqueeze_axes, int64_t axis);
+
+// Quantization modes for QuantizeLinear/DequantizeLinear.
+enum class QuantizationMode : uint8_t {
+  kUnknown,
+  kPerTensor,
+  kPerAxis,
+  kBlocked,
+};
+
+// Returns true if this optimizer supports the given quantization mode.
+constexpr bool IsSupportedQuantizationMode(QuantizationMode mode) {
+  return (mode == QuantizationMode::kPerTensor) || (mode == QuantizationMode::kPerAxis);
+}
+
+// Stores quantization info for a validated Q or DQ node.
+struct QuantizationInfo {
+  QuantizationMode mode;
+  int64_t norm_axis;  // 'normalized' axis is in the range [0, Input0Rank - 1]
+};
+
+// Returns the shape of a Q or DQ node's scale input, or std::nullopt if unable to get the shape.
+static std::optional<std::vector<int64_t>> GetQOrDQScaleShape(const api::GraphRef& graph,
+                                                              const api::NodeRef& q_or_dq_node) {
+  // Try to get the shape from ValueInfo.
+  std::string_view scale_input = q_or_dq_node.Inputs()[1];
+  const std::unique_ptr<api::ValueInfoRef> scale_value_info = graph.GetValueInfo(scale_input);
+  std::optional<std::vector<int64_t>> scale_shape = scale_value_info->Shape();
+  if (scale_shape) {
+    return scale_shape;
+  }
+
+  // ValueInfo is not available. Try to get shape from Tensor if scale is a constant.
+  const std::unique_ptr<api::TensorRef> dq_scale_constant = graph.GetConstant(scale_input);
+  if (dq_scale_constant) {
+    return dq_scale_constant->Shape();
+  }
```

Comment:
Is there a scenario where a constant does not have a NodeArg with shape info? I wouldn't have expected so, so I'm not sure if this code is useful.

Reply:
You're right. There is no scenario in which a constant would not have a valid NodeArg. The `onnxruntime::Graph` class explicitly adds NodeArgs for all constants. I removed this unnecessary logic.